File: daily-docket-pwa/js/filters.js
/* Deciding what actually counts as a commitment. */

/* Separators people type between a tag and a title. The middle dot in "WR ·"
   is U+00B7, which is easy to fat-finger, hence lenient matching. */
const SEP_CLASS = '·•‧∙・|/\\\\:;,~»>.\\u2013\\u2014\\u2212-';
const SEP_RE = new RegExp(`[${SEP_CLASS}\\s]+`, 'g');

function escapeRe(s) {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function buildPrefixMatchers(prefixes, lenient) {
  return (prefixes || [])
    .filter(Boolean)
    .map((prefix) => {
      const norm = String(prefix).normalize('NFC').trim();
      if (!norm) return null;

      if (!lenient) {
        // Exact prefix, but tolerant about how much whitespace is in it.
        const src = escapeRe(norm.replace(/\s+/g, ' ')).replace(/ /g, '\\s+');
        return { source: prefix, re: new RegExp(`^${src}`, 'i') };
      }

      // Lenient: strip the decoration down to the tag itself ("WR ·" → "WR"),
      // then require a separator or whitespace after it so "WRAP" is safe.
      const core = norm.replace(SEP_RE, '');
      if (!core) return null;
      return {
        source: prefix,
        re: new RegExp(`^${escapeRe(core)}(?=$|[${SEP_CLASS}\\s])`, 'i'),
      };
    })
    .filter(Boolean);
}

/**
 * @returns {null|{reason:string, detail?:string}} null means "keep this event"
 */
export function rejectReason(item, cfg, matchers) {
  const title = (item.title || '').normalize('NFC').trim();

  for (const m of matchers) {
    if (m.re.test(title)) return { reason: 'prefix', detail: m.source };
  }
  for (const src of cfg.ignoreTitlePatterns || []) {
    try {
      if (new RegExp(src, 'i').test(title)) return { reason: 'pattern', detail: src };
    } catch {
      /* a bad regex in config shouldn't break the docket */
    }
  }
  if (cfg.skipCancelled && item.status === 'CANCELLED') {
    return { reason: 'cancelled' };
  }
  if (cfg.skipTransparent && item.transparent) {
    // All-day items are exempt by default: Google marks birthdays and
    // milestones free, and dropping them would empty the chip row.
    if (!item.allDay || cfg.skipTransparentAllDay) return { reason: 'free' };
  }
  return null;
}

/**
 * Apply the filters to a list of expanded occurrences.
 * @returns {{kept:Array, dropped:Array}}
 */
export function applyFilters(items, cfg) {
  const matchers = buildPrefixMatchers(cfg.ignoreTitlePrefixes, cfg.lenientPrefixMatch);
  const kept = [];
  const dropped = [];
  for (const item of items) {
    const reject = rejectReason(item, cfg, matchers);
    if (reject) dropped.push({ ...item, rejected: reject });
    else kept.push(item);
  }
  return { kept, dropped };
}

/**
 * Split a day into all-day chips and timed events.
 *
 * All-day items use plain overlap, so a multi-day trip chips on every day it
 * covers. Timed events belong to the day they START on — otherwise last
 * night's 11pm call would head up tomorrow's timeline. `includeRunning` adds
 * back anything already in progress when the day began, which is what you want
 * for today and not for tomorrow.
 */
export function forDay(items, dayStart, dayEnd, { includeRunning = false } = {}) {
  const from = +dayStart;
  const to = +dayEnd;

  const allDay = items.filter(
    (i) => i.allDay && +i.start < to && Math.max(+i.end, +i.start + 1) > from,
  );

  const timed = items
    .filter((i) => {
      if (i.allDay) return false;
      if (+i.start >= from && +i.start < to) return true;
      return includeRunning && +i.start < from && +i.end > from;
    })
    .sort((a, b) => a.start - b.start || a.end - b.end);

  return { allDay, timed };
}

/** Today's timed events, sliced by the current moment. */
export function splitByNow(timed, now) {
  const ahead = [];
  const done = [];
  for (const item of timed) {
    const endsAfterNow = +item.end > +now;
    const item2 = { ...item, inProgress: +item.start <= +now && endsAfterNow };
    (endsAfterNow ? ahead : done).push(item2);
  }
  return { ahead, done };
}

File: daily-docket-ios/DailyDocket/Services/Filters.swift
import Foundation

/// Ported from js/filters.js — deciding what counts as a real commitment.
enum Filters {
    struct PrefixMatcher {
        let source: String
        let regex: NSRegularExpression
    }

    enum RejectReason {
        case prefix(String)
        case pattern(String)
        case cancelled
        case free
    }

    /// Separators people type between a tag and a title (mirrors SEP_CLASS in
    /// js/filters.js: middle dot, bullet, various dashes, punctuation, ~).
    private static let sepClassPattern = #"[·•‧∙・|/\\:;,~»>.–—−\s-]+"#
    private static let sepRegex = try! NSRegularExpression(pattern: sepClassPattern)

    private static func stripSeparators(_ s: String) -> String {
        let range = NSRange(s.startIndex..., in: s)
        return sepRegex.stringByReplacingMatches(in: s, range: range, withTemplate: "")
    }

    static func buildPrefixMatchers(_ prefixes: [String], lenient: Bool) -> [PrefixMatcher] {
        var out: [PrefixMatcher] = []
        for prefix in prefixes {
            let norm = prefix.precomposedStringWithCanonicalMapping.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !norm.isEmpty else { continue }

            if !lenient {
                let collapsedRange = NSRange(norm.startIndex..., in: norm)
                let collapser = try! NSRegularExpression(pattern: #"\s+"#)
                let collapsed = collapser.stringByReplacingMatches(in: norm, range: collapsedRange, withTemplate: " ")
                let escaped = NSRegularExpression.escapedPattern(for: collapsed).replacingOccurrences(of: " ", with: #"\s+"#)
                guard let re = try? NSRegularExpression(pattern: "^" + escaped, options: [.caseInsensitive]) else { continue }
                out.append(PrefixMatcher(source: prefix, regex: re))
                continue
            }

            let core = stripSeparators(norm)
            guard !core.isEmpty else { continue }
            let escapedCore = NSRegularExpression.escapedPattern(for: core)
            let lookahead = #"(?=$|[·•‧∙・|/\\:;,~»>.–—−\s-])"#
            guard let re = try? NSRegularExpression(pattern: "^" + escapedCore + lookahead, options: [.caseInsensitive]) else { continue }
            out.append(PrefixMatcher(source: prefix, regex: re))
        }
        return out
    }

    private static func matches(_ regex: NSRegularExpression, _ s: String) -> Bool {
        regex.firstMatch(in: s, range: NSRange(s.startIndex..., in: s)) != nil
    }

    static func rejectReason(_ item: EventItem, cfg: AppConfig, matchers: [PrefixMatcher]) -> RejectReason? {
        let title = item.title.precomposedStringWithCanonicalMapping.trimmingCharacters(in: .whitespacesAndNewlines)

        for m in matchers where matches(m.regex, title) {
            return .prefix(m.source)
        }
        for src in cfg.ignoreTitlePatterns {
            if let re = try? NSRegularExpression(pattern: src, options: [.caseInsensitive]), matches(re, title) {
                return .pattern(src)
            }
        }
        if cfg.skipCancelled, item.status == "CANCELLED" { return .cancelled }
        if cfg.skipTransparent, item.transparent {
            if !item.allDay || cfg.skipTransparentAllDay { return .free }
        }
        return nil
    }

    static func apply(_ items: [EventItem], cfg: AppConfig) -> (kept: [EventItem], dropped: [EventItem]) {
        let matchers = buildPrefixMatchers(cfg.ignoreTitlePrefixes, lenient: cfg.lenientPrefixMatch)
        var kept: [EventItem] = []
        var dropped: [EventItem] = []
        for item in items {
            if rejectReason(item, cfg: cfg, matchers: matchers) != nil {
                dropped.append(item)
            } else {
                kept.append(item)
            }
        }
        return (kept, dropped)
    }

    /// All-day items overlap the day; timed events belong to the day they
    /// start on, plus already-running ones when `includeRunning` is set.
    static func forDay(_ items: [EventItem], dayStart: Date, dayEnd: Date, includeRunning: Bool = false) -> (allDay: [EventItem], timed: [EventItem]) {
        let allDay = items.filter { $0.allDay && $0.start < dayEnd && max($0.end, $0.start.addingTimeInterval(1)) > dayStart }
        let timed = items
            .filter { item in
                guard !item.allDay else { return false }
                if item.start >= dayStart, item.start < dayEnd { return true }
                return includeRunning && item.start < dayStart && item.end > dayStart
            }
            .sorted { $0.start != $1.start ? $0.start < $1.start : $0.end < $1.end }
        return (allDay, timed)
    }

    static func splitByNow(_ timed: [EventItem], now: Date) -> (ahead: [EventItem], done: [EventItem]) {
        var ahead: [EventItem] = []
        var done: [EventItem] = []
        for item in timed {
            if item.end > now { ahead.append(item) } else { done.append(item) }
        }
        return (ahead, done)
    }
}
